[SYCL-PTX] Update documentation for the CUDA backend. (#1820)
intel/llvm: sycl/doc/CompilerAndRuntimeDesign.md

`libdevice.bc` (provided by the CUDA SDK) and `libspirv-nvptx64--nvidiacl.bc`
(built by the libclc project).

##### Device code post-link step

> During the "PTX target precessing" in the device linking step [Device code post-link step](#device-code-post-link-step), the llvm bitcode objects for the CUDA target are linked together alongside `libspirv-nvptx64--nvidiacl.bc` and `libdevice.bc`, compiled to PTX

Reviewer: ```suggestion
During the "PTX target processing" in the device linking step [Device code post-link step](#device-code-post-link-step), the llvm bitcode objects for the CUDA target are linked together alongside `libspirv-nvptx64--nvidiacl.bc` and `libdevice.bc`, compiled to PTX
```

Reviewer: Please format to 80 chars lines.

Reviewer 2: on the picture I can see "libspirv" and "libdevice.bc" passed to "Clang"
1) Looks like "libspirv" should be "libspirv.bc"
2) Looks like "clang" is really PTX back-end - I suggest to replace "Clang" with "PTX backend" then, similarly to the generic DeviceLinkAndWrap terminology

Also does PTX BE accept multiple bitcode files or there is llvm-link before it? If the latter, I'd suggest to reflect that on the picture.

Author: I applied your comments on the picture.

It is llvm-link before, but it is clearly shown in the DeviceLinkAndWrap, the picture is a split of the "PTX target processing" box. I added reminder on the input box that it is linked.

On the runtime side, when setting local memory arguments, the CUDA PI
implementation will internally set the argument as the offset with respect to
the accumulated size of used local memory. This approach preserves the existing
PI interface.

##### Global offset support

The CUDA API does not natively support the global offset parameter expected by the SYCL.

> In order to emulate this and make generated kernel compliant, an intrinsic `llvm.nvvm.implicit.offset` (clang builtin `__builtin_ptx_implicit_offset`) was introduce materializing the use of this implicit parameter for the NVPTX backend. The intrinsic return a pointer to `i32` referring to a 3 elements array.

Reviewer: Please format to 80 chars lines

> In order to emulate this and make generated kernel compliant, an intrinsic `llvm.nvvm.implicit.offset` (clang builtin `__builtin_ptx_implicit_offset`) was introduce materializing the use of this implicit parameter for the NVPTX backend. The intrinsic return a pointer to `i32` referring to a 3 elements array.

Reviewer 2: ```suggestion
In order to emulate this and make generated kernel compliant, an intrinsic `llvm.nvvm.implicit.offset` (clang builtin `__builtin_ptx_implicit_offset`) was introduced materializing the use of this implicit parameter for the NVPTX backend. The intrinsic returns a pointer to `i32` referring to a 3 elements array.
```

Each non-kernel function reaching the implicit offset intrinsic in the call graph is augmented with an extra implicit parameter of type pointer to `i32`.

> Each kernel calling one of these functions using this intrinsic are cloned:

Reviewer 2: ```suggestion
Kernels calling one of these functions using this intrinsic are cloned:
```

> - the original kernel initialize an array of 3 `i32` to 0 and passes the pointer to this array to each function with the implicit parameter;

Reviewer 2: ```suggestion
- the original kernel initializes an array of 3 `i32` to 0 and passes the pointer to this array to each function with the implicit parameter;
```

- the cloned function type is augmented with an implicit parameter of type array of 3 `i32`. The pointer to this array is then passed each function with the implicit parameter.

> The runtime will query both kernel and call the appropriate one based on the following logic:

Reviewer 2: ```suggestion
The runtime will query both kernels and call the appropriate one based on the following logic:
```
